File: frontend/lib/models/dtos/user.dto.dart
import 'package:flutter/material.dart';
import 'package:foodz/endpoints/base/json.serializable.dart';
import 'package:foodz/models/dtos/pagination.options.dto.dart';
import 'package:foodz/util/util.dart';

import '../status.model.dart';

class GetUserInfoDto implements JsonSerializable {
  String userId;
  GetUserInfoDto({@required this.userId});

  Map<String, dynamic> toJson() {
    return {"userId": this.userId};
  }
}

class GetUsersDto implements JsonSerializable {
  final PaginationOptions paginationOptions;
  GetUsersDto({this.paginationOptions});

  @override
  Map<String, dynamic> toJson() {
    return onlyDefinedKeys(
        {"paginationOptions": this.paginationOptions?.toJson()});
  }
}

class UpdateUserDto implements JsonSerializable {
  final String id;
  final String firstName;
  final String lastName;
  final Status status;
  UpdateUserDto(
      {@required this.id, this.firstName, this.lastName, this.status});

  @override
  Map<String, dynamic> toJson() {
    return onlyDefinedKeys({
      'id': id,
      'firstName': firstName,
      'lastName': lastName,
      'status': status.string,
    });
  }
}

class DeleteUserDto implements JsonSerializable {
  final String id;
  DeleteUserDto({@required this.id});

  @override
  Map<String, dynamic> toJson() {
    return {'id': id};
  }
}

File: backend/src/dtos/user.dto.ts
import { Status } from '@interfaces/status.interface';
import { UserRole } from '@interfaces/user.interface';
import { IsEmail, IsEnum, IsNotEmpty, IsOptional, IsString } from 'class-validator';
import { PaginationOptions } from './pagination.options.dto';

export class RegisterUserDto {
  @IsEmail()
  public email: string;

  @IsString()
  @IsNotEmpty()
  public password: string;

  @IsEnum(UserRole)
  public role: UserRole;

  @IsString()
  @IsNotEmpty()
  public firstName: string;

  @IsString()
  @IsNotEmpty()
  public lastName: string;
}

export class LoginUserDto {
  @IsEmail()
  public email: string;

  @IsString()
  @IsNotEmpty()
  public password: string;
}

export class GetUserInfoDto {
  @IsString()
  @IsNotEmpty()
  public id: string;
}

export class RefreshTokenDto {
  @IsString()
  @IsNotEmpty()
  public token: string;
}

export class GetUsersDto {
  @IsOptional()
  paginationOptions: PaginationOptions;
}

export class UpdateUserDto {
  @IsString()
  @IsNotEmpty()
  public id: string;

  @IsOptional()
  @IsString()
  @IsNotEmpty()
  public firstName: string;

  @IsOptional()
  @IsString()
  @IsNotEmpty()
  public lastName: string;

  @IsOptional()
  @IsEnum(Status)
  public status: Status;
}

export class DeleteUserDto {
  @IsString()
  @IsNotEmpty()
  public id: String;
}
